### DISCUSSION TITLE
What is the best way to work with utf-8 encoded string

### DISCUSSION THREAD

**Discussion by aaa006bd** [kjur/jsrsasign#584] [Q&A] [answered]
Hi,
I have an issue regarding creating signatures from strings containing special characters e.g. 'Zeyestraße'. 
I was trying to create a messageDigest using 
```
 const messageDigest = new jsrsasign.KJUR.crypto.MessageDigest({
    alg: 'sha256',
    prov: 'cryptojs',
  });
const dataString = canonicalize(data); // https://github.com/snowyu/json-canonicalize.ts/blob/master/src/canonicalize.ts
messageDigest.updateHex(jsrsasign.utf8tohex(dataString));
```
Here I need to use updateHex because the datastring might be constructed from a Json object as well. 

The way I am doing it is not working because the signature is not created properly (UTF-8 content must be signed in the correct UTF-8 encoding). Is there anything that I am missing? or What should be the way to do it.

**Comment by kjur** [comment, marked as answer]
Jsrsasign works with UTF-8 non ascii characters. Here is a sample input JSON file "t1.json" contents (no new line in end of file):
```JavaScript
{
  "text": "Zeyestraße"
}
```
SHA256 hash value of this file will be:
```
$ openssl sha256 t1.json
SHA256(t1.json)= 6bbaebd8da1853f65298b72a970c3a44b9d4d2bdf3299330968441efe8fd7c72
```

Then here is hash calculation sample of jsrsasign with updateString method:
```JavaScript
const jsrsasign = require("jsrsasign");
const fs = require("fs");
const s = fs.readFileSync("t1.json", "utf8");
let md = new jsrsasign.KJUR.crypto.MessageDigest({alg: "sha256"});
md.updateString(s);
console.log(md.digest());
```

Here is another sample with updateHex method:
```JavaScript
const jsrsasign = require("jsrsasign");
const fs = require("fs");
const s = fs.readFileSync("t1.json", "utf8");
let md = new jsrsasign.KJUR.crypto.MessageDigest({alg: "sha256"});
md.updateHex(jsrsasign.utf8tohex(s));
console.log(md.digest());
```

Both script will show the same correct hash value:
```
6bbaebd8da1853f65298b72a970c3a44b9d4d2bdf3299330968441efe8fd7c72
```

I hope these help you.
